### PR TITLE
feat(esl-utils): update `object#get` and `object#set` with extended syntax support

### PR DESCRIPTION
**Pull request by ala-n**
Created extended version of `set` method #831 
The `get` extended method was created as well.
Structure docs and tests updated for `misc/object`.

Closes: #831

### CODE REVIEW

File: src/modules/esl-utils/misc/object.ts [hunk 1]
```diff
@@ -89,8 +89,34 @@ export const set = (target: any, path: string, value: any): void => {
...
+ * @param value - value of property
+ */
+export const setExt = (target: any, path: string, value: any): void => {
+  // TODO: micro-optimization if possible for the following lines
+  // remove initial bracket
+  path = (path || '').replace(/^\[/, '');
+  // split using special strategy:
+  // '[' join to the previous key to identify array consuming on the next part
+  // ']' is part of the current key to identify array index operation
+  const parts = path.replace(/(^|.)\[/g, '$1[.').split('.');
```
Comment (dshovchko) [inline src/modules/esl-utils/misc/object.ts:111]:
```suggestion
export const setExt = (target: any, path: string = '', value: any): void => {
  // TODO: micro-optimization if possible for the following lines
  // remove initial bracket
  // split using special strategy:
  // '[' join to the previous key to identify array consuming on the next part
  // ']' is part of the current key to identify array index operation
  const parts = path.replace(/^\[/, '').replace(/(^|.)\[/g, '$1[.').split('.');
```

**Review by dshovchko — APPROVED** [review]

File: src/modules/esl-utils/misc/object.ts [hunk 2]
```diff
@@ -111,6 +96,75 @@ export const get = (data: any, path: string, defaultValue?: any): any => {
...
+  while (match = matcher.exec(path)) {
+    const [, key1, key2, index] = match;
+    if (index !== undefined && (!index || !isNaN(+index))) {
+      parts.push({key: index, isIndex: true});
+    } else {
+      parts.push({key: key1 || key2 || index || ''});
+    }
+  }
+  return parts;
+};
```
Comment (ala-n) [inline src/modules/esl-utils/misc/object.ts:135]: @exadel-inc/esl-core-team , 
this is the last thing that currently under optimization, everything else is ready to review

**Comment by ala-n** [comment]
Sorry, not ready at all for now.
Total API inconsistency that I have fixed a commit above still doesn't look good from an API and performance point of view.
Need I think we need once more regarding the final API of the #831 request.
@exadel-inc/esl-core-team

File: src/modules/esl-utils/misc/object/copy.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  });
+  return result;
+}
+
+/** Makes a flat copy without undefined keys */
+export function copyDefinedKeys<T>(obj?: T): Partial<T> {
+  return copy(obj || {}, (key, value) => value !== void 0);
+}
+
+/** Omit copying provided properties from object */
```
Comment (julia-murashko) [inline src/modules/esl-utils/misc/object/copy.ts:18]:
```suggestion
/** Omits copying provided properties from object */
```

File: src/modules/esl-utils/misc/object/path.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  // /** Key should produce array is not exists */
+  isIndexed?: boolean;
+};
+/** Key definition for {@link getBy} or {@link setBy} */
+export type PathKey = PathKeyDef | string | number;
+
+/** @returns PathKeyDef from the PathDef */
+const toKeyDef = (key: PathKey): PathKeyDef => typeof key === 'object' ? key : {key};
+
+/** Parse path to full {@link PathKeyDef} array */
```
Comment (julia-murashko) [inline src/modules/esl-utils/misc/object/path.ts:18]:
```suggestion
/** Parses path to full {@link PathKeyDef} array */
```

File: src/modules/esl-utils/misc/object/path.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+export const parseKeys = (path: string | PathKey[]): PathKeyDef[] => {
+  if (Array.isArray(path)) return path.map(toKeyDef);
+  return parseKeysPath(path || '.');
+};
+
+/** Parse string path to full {@link PathKeyDef} array */
```
Comment (julia-murashko) [inline src/modules/esl-utils/misc/object/path.ts:24]:
```suggestion
/** Parses string path to full {@link PathKeyDef} array */
```

File: src/modules/esl-utils/misc/object/path.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  const keys = parseKeys(path);
+  const result = keys.reduce((curr: any, {key}: PathKeyDef) => {
+    if (isObjectLike(curr)) return curr[key];
+    return undefined;
+  }, data);
+  return typeof result === 'undefined' ? defaultValue : result;
+};
+
+/**
+ * Get object property using "path" with a keys separated by `.`
```
Comment (julia-murashko) [inline src/modules/esl-utils/misc/object/path.ts:75]:
```suggestion
 * Gets object property using "path" with a keys separated by `.`
```

File: src/modules/esl-utils/misc/object/path.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+ * @see getBy
+ */
+export const get = (data: any, path: string, defaultValue?: any): any => getBy(data, (path || '').split('.'), defaultValue);
+
+/**
+ * Set object property using "path" key
```
Comment (julia-murashko) [inline src/modules/esl-utils/misc/object/path.ts:81]:
```suggestion
 * Sets object property using "path" key
```

File: src/modules/esl-utils/misc/object/path.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    if (isIndex && !key) key = cur.length || 0; // a[] only
+    if (pos !== depth && isObjectLike(cur[key])) return cur[key]; // key already presented
+    if (isIndexed === undefined && pos !== depth) isIndexed = keys[pos + 1].isIndex;
+    return cur[key] = (pos === depth) ? value : (isIndexed ? [] : {});
+  }, target);
+  return target;
+};
+
+/**
+ * Set object property using "path" with a keys separated by `.`
```
Comment (julia-murashko) [inline src/modules/esl-utils/misc/object/path.ts:111]:
```suggestion
 * Sets object property using "path" with a keys separated by `.`
```

**Review by julia-murashko — APPROVED** [review]

File: src/modules/esl-utils/misc/object/path.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
+import {isObjectLike} from './types';
+
+/** Full key definition for {@link getBy} or {@link setBy} */
+export type PathKeyDef = {
+  /** Key name */
+  key: string | number;
+  /** Key represents collection index */
+  isIndex?: boolean;
+  // /** Key should produce array is not exists */
```
Comment (yadamskaya) [inline src/modules/esl-utils/misc/object/path.ts:9]:
```suggestion
  /** Key should produce array is not exists */
```

**Review by yadamskaya — APPROVED** [review]

**Review by abarmina — APPROVED** [review]

**Review by NastaLeo — APPROVED** [review]

**Pull request merged**